tree-building : implement exercise

clapmyhands opened the pull request:
resolves #741

m-a-ge commented on exercises/tree-building/example.py:
```diff
@@ -0,0 +1,44 @@
+
+def validateRecord(record):
+    if record.equal_id() and record.record_id != 0:
+        raise ValueError
```
Should probably provide the cause of the exception

m-a-ge commented on exercises/tree-building/example.py:
```diff
@@ -0,0 +1,44 @@
+    if record.equal_id() and record.record_id != 0:
+        raise ValueError
+    elif (not record.equal_id()) and record.parent_id >= record.record_id:
+        raise ValueError
```
the same

m-a-ge commented on exercises/tree-building/example.py:
```diff
@@ -0,0 +1,44 @@
+def validateRecord(record):
+    if record.equal_id() and record.record_id != 0:
+        raise ValueError
+    elif (not record.equal_id()) and record.parent_id >= record.record_id:
```
redundant parentheses

clapmyhands commented:
@m-a-ge I've change it according to the review. Please check again and let me know if there's anything else to change 😄 .

cmccandless commented on exercises/tree-building/tree_building_test.py:
```diff
@@ -0,0 +1,178 @@
+
+    def assert_node_is_branch(self, node, node_id, children_count):
+        self.assertEqual(node.node_id, node_id)
+        self.assertFalse(len(node.children) == 0)
```
`assertNotEqual(len(node.children), 0)` would be clearer.

cmccandless commented on exercises/tree-building/tree_building_test.py:
```diff
@@ -0,0 +1,178 @@
+
+    def assert_node_is_leaf(self, node, node_id):
+        self.assertEqual(node.node_id, node_id)
+        self.assertTrue(len(node.children) == 0)
```
`assertEqual(len(node.children), 0)` would be clearer.

clapmyhands commented:
@cmccandless I've changed according to your review. Also I've been thinking, do I need to put some reason on the test where a `ValueError` need to be raised?

cmccandless commented:
@clapmyhands If I understand you correctly, do you mean something like this?
``` Python
with self.assertRaisesRegexp(ValueError, 'No root node'):
    BuildTree(records)
```

If that's what you mean, that would be a good idea :+1:

clapmyhands commented:
@cmccandless I don't think it's possible. At least in my current code it can't since the checking of 'no root node' is the same as checking of 'non continuous'

reason being:
```python
raise ValueError
```
 can cover multiple failing cases.

I was thinking more in line of comments like so:
```python
# no root node
with self.assertRaises(ValueError):
            BuildTree(records)
```

the name of the test might already be self-explanatory but for some tests such as
```python
def test_cycle_directly(self):
```
it might be good to have further explanation on why that gives an error.

p.s: what do you think about changing
```python
self.assertIs(root, None)
```
to
```python
self.assertIsNone(root)
```

clapmyhands commented:
@cmccandless I've changed it to use `self.assertIsNone` and added comment where I think is needed. Please review once more and let me know if there's anything else I need to change 😄

cmccandless commented:
@clapmyhands I've left a couple comments. Just some consistency stuff.

clapmyhands commented:
@cmccandless I've added comment for consistency.

m-a-ge commented:
Thanks for working on this, @clapmyhands!

clapmyhands commented:
@m-a-ge @cmccandless Thanks for being patient with me through it all. New contributor and hopefully a lasting one 😄
